python3Packages.orange3: init at 3.34.0

totoroot opened the pull request:
###### Description of changes

Adds the "Orange Data Mining" application and all its dependencies as Python modules.
This closes issue #208574. More information about the application can be found in the issue's description. 

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

totoroot committed:
anyqt: init at 0.2.0

totoroot committed:
baycomp: init at 1.0.2

totoroot committed:
opentsne: init at 0.6.2

totoroot committed:
serverfiles: init at 0.3.0

totoroot committed:
qasync: init at 0.19.0

totoroot committed:
orange-canvas-core: init at 0.1.28

totoroot committed:
orange-widget-base: init at 4.19.0

totoroot committed:
orange3: init at 3.34.0

risicle commented on pkgs/development/python-modules/baycomp/default.nix:
```diff
@@ -0,0 +1,41 @@
+  };
+
+  buildInputs = [ ];
+  nativeBuildInputs = [ setuptools setuptools_scm];
```
`setuptools-scm`

totoroot replied:
Thanks, fixed

fabaff reviewed (commented):
I only mentioned the first occurrence of an issue.

risicle commented on pkgs/development/python-modules/anyqt/default.nix:
```diff
@@ -0,0 +1,33 @@
+    description = "A PyQt/PySide compatibility layer";
+    homepage = "https://anyqt.readthedocs.io";
+    changelog = "https://github.com/ales-erjavec/anyqt/releases/tag/${version}";
+    downloadPage = changelog;
```
To be able to do this you need to use the `rec` keyword before this attrset's opening brace.

totoroot replied:
Thanks, but that didn't work. ofborg-eval is then running into issues since `maintainers` gets redefined.
I've just removed the `downloadPage` attribute if it was just a copy of `changelog` since it is not strictly necessary.

totoroot commented:
> I'll echo that adding 7 new packages without tests enabled doesn't make for happy maintenance.

Sure, makes sense. I'm working on getting the tests working :v:

totoroot commented:
I cannot get the tests to work for `qasync`. Do I need some additional `checkInputs` besides `pyqt5`?
    
    Check whether the following modules can be imported: qasync
    pytestcachePhase
    pytestRemoveBytecodePhase
    pytestCheckPhase
    Executing pytestCheckPhase
    ============================= test session starts ==============================
    platform linux -- Python 3.10.9, pytest-7.1.3, pluggy-1.0.0
    rootdir: /build/source
    collected 44 items
    
    tests/test_qeventloop.py Fatal Python error: Aborted
    
    Current thread 0x00007ffff7e7f740 (most recent call first):
      File "/build/source/tests/conftest.py", line 26 in application
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 900 in call_fixture_func
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 1121 in pytest_fixture_setup
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 1067 in execute
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 669 in _compute_fixture_value
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 583 in _get_active_fixturedef
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 564 in getfixturevalue
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/fixtures.py", line 551 in _fillfixtures
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/python.py", line 1764 in setup
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 491 in setup
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 154 in pytest_runtest_setup
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 259 in <lambda>
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 338 in from_call
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 258 in call_runtest_hook
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 219 in call_and_report
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 124 in runtestprotocol
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/runner.py", line 111 in pytest_runtest_protocol
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/main.py", line 347 in pytest_runtestloop
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/main.py", line 322 in _main
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/main.py", line 268 in wrap_session
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/main.py", line 315 in pytest_cmdline_main
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_callers.py", line 39 in _multicall
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
      File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/config/__init__.py", line 164 in main
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/config/__init__.py", line 187 in console_main
      File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/pytest/__main__.py", line 5 in <module>
      File "/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/runpy.py", line 86 in _run_code
      File "/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/runpy.py", line 196 in _run_module_as_main
    
    Extension modules: PyQt5.QtCore, PyQt5.QtGui, PyQt5.QtWidgets (total: 3)
    /nix/store/ppxrfnv4b3pvcxpncipdj6z5v6rsd72n-pytest-check-hook/nix-support/setup-hook: line 53:   127 Aborted                 (core dumped) /nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/bin/python3.10 -m pytest
    error: builder for '/nix/store/bsrxjsawzkbcccwp60yb1c86vhj6rxly-python3.10-qasync-0.19.0.drv' failed with exit code 134;
           last 10 log lines:
           >   File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_manager.py", line 80 in _hookexec
           >   File "/nix/store/2xmrdwg8q9yfryq1297pv6kqx9h0wgya-python3.10-pluggy-1.0.0/lib/python3.10/site-packages/pluggy/_hooks.py", line 265 in __call__
           >   File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/config/__init__.py", line 164 in main
           >   File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/_pytest/config/__init__.py", line 187 in console_main
           >   File "/nix/store/96mm3zbb0f3gdxd2fv7ml4xsskxw90q8-python3.10-pytest-7.1.3/lib/python3.10/site-packages/pytest/__main__.py", line 5 in <module>
           >   File "/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/runpy.py", line 86 in _run_code
           >   File "/nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/lib/python3.10/runpy.py", line 196 in _run_module_as_main
           >
           > Extension modules: PyQt5.QtCore, PyQt5.QtGui, PyQt5.QtWidgets (total: 3)
           > /nix/store/ppxrfnv4b3pvcxpncipdj6z5v6rsd72n-pytest-check-hook/nix-support/setup-hook: line 53:   127 Aborted                 (core dumped) /nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/bin/python3.10 -m pytest
           For full logs, run 'nix log /nix/store/bsrxjsawzkbcccwp60yb1c86vhj6rxly-python3.10-qasync-0.19.0.drv'.

totoroot commented:
Thank you very much, @fabaff and @risicle! :pray:
I feel like I've already learned a lot from your reviews.
I tried fixing the issues according to your feedback and have resolved the comments, that I've dealt with.

**Tests are still not enabled for all added packages**
I've enabled checks/tests for `baycomp` and `serverfiles` without problems.
The checks for `anyqt` and `qasync` fail due to some issues with pyqt5 modules and the tests for `opentsne` seem to require some archive that  I do not know how to pass.

risicle commented on pkgs/development/python-modules/opentsne/default.nix:
```diff
@@ -0,0 +1,50 @@
+  ];
+
+  # No such file or directory: 'data/macosko_2015.pkl.gz'
+  doCheck = false;
```
```
  checkInputs = [ pytestCheckHook ];
  preCheck = let
    macosko2015 = fetchurl {
      url = "https://file.biolab.si/opentsne/benchmark/macosko_2015.pkl.gz";
      hash = "sha256-WFhQwfdmtaNRR1olTj6EWlWceajoL/12WA+t1HzDrvA=";
    };
  in ''
    mkdir -p data
    ln -s ${macosko2015} data/macosko_2015.pkl.gz

    # ensure tests find installed package, not source dir
    mv openTSNE openTSNE.hidden
  '';
  disabledTests = lib.optionals (!stdenv.hostPlatform.isx86_64) [
    # overly precise comparison against reference
    # presumably calculated on x86
    "test_iris"
  ];
```

Obviously you'll need to provide `fetchurl`, `stdenv` and `pytestCheckHook` via the args.

risicle commented on pkgs/development/python-modules/orange-widget-base/default.nix:
```diff
@@ -0,0 +1,53 @@
+    pyqt5
+    pyqtgraph
+    typing-extensions
+  ];
```
```suggestion
  ] ++ lib.optionals stdenv.isDarwin [
    appnope
  ];
```

Needed on darwin.

totoroot committed:
opentsne: add tests

totoroot committed:
orange3: add tests for dependencies

anyqt: add some tests
orange-canvas-core: add some tests
orange-widget-base: add some tests
qasync: add some tests

totoroot committed:
orange3: try to make tests work

totoroot commented:
According to [this section in the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#qt-runtime-dependencies) I should be able to get rid of the following error by correctly managing the wrapping of the application.

However, the tests for `orange3` still fail:

     > qt.plugin: Could not find the Qt platform plugin "xcb" in ""
     > This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
     >
     > Available platform plugins are: wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx.
     >
     > /nix/store/dvgjig1ny83n264fgdkrc3y1hiw55cq7-setuptools-check-hook/nix-support/setup-hook: line 4:  2190 Aborted                 (core dumped) /nix/store/al6g1zbk8li6p8mcyp0h60d08jaahf8c-python3-3.10.9/bin/python3.10 nix_run_setup test

risicle commented:
It's not always possible to get all test suites running, especially if they use gui functionality.

risicle commented on pkgs/development/python-modules/orange-canvas-core/default.nix:
```diff
@@ -0,0 +1,79 @@
+    pytest -k test_qobjref.py
+    pytest -k test_shtools.py
+    pytest -k test_utils.py
+  '';
```
```
  pytestFlagsArray = [
    "orangecanvas/registry/tests/test_base.py"
    "orangecanvas/registry/tests/test_discovery.py"
    "orangecanvas/scheme/tests/test_annotations.py"
    "orangecanvas/scheme/tests/test_signalmanager.py"
    "orangecanvas/utils/tests/test_after_exit.py"
    "orangecanvas/utils/tests/test_graph.py"
    "orangecanvas/utils/tests/test_markup.py"
    "orangecanvas/utils/tests/test_qinvoke.py"
    "orangecanvas/utils/tests/test_qobjref.py"
    "orangecanvas/utils/tests/test_shtools.py"
    "orangecanvas/canvas/items/tests/test_utils.py"
    "orangecanvas/utils/tests/test_utils.py"
  ];
```

Otherwise you're not really using the `pytestCheckHook`, instead you're replacing the whole `checkPhase`

risicle commented on pkgs/development/python-modules/anyqt/default.nix:
```diff
@@ -0,0 +1,50 @@
+  ];
+
+  # All of these fail because Qt modules cannot be imported
+  disabledTests = [
```
Better to use `disabledTestPaths`

risicle commented on pkgs/development/python-modules/qasync/default.nix:
```diff
@@ -0,0 +1,56 @@
+  # FIXME Cannot get `test_qeventloop.py` to work, perhaps due to bad OS detection in test
+  # https://github.com/CabbageDevelopment/qasync/blob/master/tests/test_qeventloop.py
+  checkPhase = ''
+    pytest -k 'test_qthreadexec.py'
```
```
pytestFlagsArray = [
  "tests/test_qthreadexec.py"
];
```

risicle commented:
On macos 10.15 `orange3` fails with:

```
test_transformation (test_transformation.TestTransformEquality) ... ok
qt.qpa.plugin: Could not find the Qt platform plugin "cocoa" in ""
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

/nix/store/5yv43761qa1mffcx7n30167n1q9917my-setuptools-check-hook/nix-support/setup-hook: line 4: 11886 Abort trap: 6           /nix/store/fbll01zhyfffmshwrrzphmbscmnimn3f-python3-3.10.9/bin/python3.10 nix_run_setup test
```

risicle commented on pkgs/development/python-modules/orange3/default.nix:
```diff
@@ -0,0 +1,97 @@
+  dontWrapQtApps = true;
+
+  disabledTests = [
+    ""
```
Also I don't think `disabledTests` does anything without `pytestCheckHook`

totoroot commented:
I cannot get the tests to work. I have to give this another shot in a few days.

hervyqa commented:
New [Release 3.35.0](https://github.com/biolab/orange3/releases/tag/3.35.0). Anyone still working on this?

totoroot commented:
Closing this in favour of #241537, which supersedes this PR.
